### PR TITLE
Refactor `grad_norm` function

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Part of #9287

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: tests/models/test_grad_norm.py [hunk 1]
```diff
@@ -105,6 +105,6 @@ def test_grad_tracking_interval(tmpdir, log_every_n_steps):
         # logging on n steps + 1 epochs
         assert len(grad_norm_dicts) == expected + 1
         # check all metrics derived from steps have the same keys
-        assert all(grad_norm_dicts[0].keys() == g.keys() for g in grad_norm_dicts[:-1])
+        assert all(grad_norm_dicts[0].keys() == g.keys() for g in grad_norm_dicts[1:-1])
```
Comment (carmocca) [inline tests/models/test_grad_norm.py:108]: Improvement: No need to check the first element with itself.

File: pytorch_lightning/utilities/grads.py [hunk 2]
```diff
@@ -35,18 +35,13 @@ def grad_norm(module: Module, norm_type: Union[float, int, str]) -> Dict[str, fl
             as a single vector.
     """
     norm_type = float(norm_type)
-
-    norms, all_norms = {}, []
```
Comment (carmocca) [inline pytorch_lightning/utilities/grads.py:39]: Improvemenet: No need to track separate objects

File: pytorch_lightning/utilities/grads.py [hunk 3]
```diff
@@ -35,18 +35,13 @@ def grad_norm(module: Module, norm_type: Union[float, int, str]) -> Dict[str, fl
...
-    for name, p in module.named_parameters():
-        if p.grad is None:
-            continue
-
-        param_norm = float(p.grad.data.norm(norm_type))
```
Comment (carmocca) [inline pytorch_lightning/utilities/grads.py:44]: Improvement: Use `.item()` over `float()`

File: pytorch_lightning/utilities/grads.py [hunk 4]
```diff
@@ -35,18 +35,13 @@ def grad_norm(module: Module, norm_type: Union[float, int, str]) -> Dict[str, fl
...
-        norms[f"grad_{norm_type}_norm_{name}"] = round(param_norm, 4)
-
-        all_norms.append(param_norm)
-
-    total_norm = float(torch.tensor(all_norms).norm(norm_type))
-    norms[f"grad_{norm_type}_norm_total"] = round(total_norm, 4)
```
Comment (carmocca) [inline pytorch_lightning/utilities/grads.py:50]: Improvement: `torch.tensor([]).norm()` is equal to `tensor(0.)` so when there are no gradients, this code stills adds the key for total norm (0).

File: pytorch_lightning/utilities/grads.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/utilities/grads.py:44]: `.item()` will trigger a device to host copy which would incur a perf hit

Comment (carmocca) [reply pytorch_lightning/utilities/grads.py:44]: Same as for `float()`, but `.item()` does not impose the type conversion

**Review by tchaton — APPROVED** [review]
LGMT !

**Review by awaelchli — APPROVED** [review]
@carmocca while looking at this review I noticed we don't have unit tests for this function. 
made a PR here and discovered we don't handle L0 properly #9765

**Review by ananthsub — APPROVED** [review]
